third_party/move/move-model/src/ast.rs

```diff
@@ -598,7 +640,20 @@ pub enum ExpData {
     /// Represents an invocation of a function value, as a lambda.
     Invoke(NodeId, Exp, Vec<Exp>),
     /// Represents a lambda.
-    Lambda(NodeId, Pattern, Exp),
+    Lambda(NodeId, Pattern, Exp, LambdaCaptureKind, AbilitySet),
+    /// Represents a reference to a Move Function as a function value.
+    MoveFunctionExp(NodeId, ModuleId, FunId),
```

Comment:
We have 2 distinct expressions we need to build:
- create a function value from a defined function by `(ModuleId, FunId)`
- create a function value from a function value and 0 or more parameters
Your `Closure` operation doesn't handle the latter case, as the "function value" need not be a reference to a defined function (`ModuleId, FunId`) but can be an arbitrary expression.

If `f` in `Operation::Closure(f, mask)` can refer to an expression, then we have bigger consistency problems.

We might conceivably have `Operation::Closure(mask)` with the function value as the first argument to the `Call` expression, though it's a bit irregular.

OTOH, we might only allow directly creating a closure based on a Move function.  This is more limiting, in that `|x| f(x)` can only be used in MVP if `f` is a user function, not if it's a function value.  I guess perhaps that is what you had in mind.  I prefer more symmetry.

Reply:
I've minimized the changes to `ExpData` by adding `Function` as a `Value`, and instead of `Closure` or `Curry` I have an `Operation` called `Bind`, which takes a function expr and the list of parameters, with the mask as a field in the variant: `Operation::Bind(mask)`.

Hopefully this is more to your liking, although since i decided to call it `Bind` I realized we have `Bind` in a few places.  Perhaps I can rename it as `EarlyBind` or something.